### PR TITLE
distsql: make CopStream available

### LINKED ISSUE #11317: distsql: make CopStream available
## Feature Request

**Is your feature request related to a problem? Please describe:**

The streaming feature in `distsql` is disabled by default for now, we should consider enabling it by default.

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

Close #11317 

### Check List 

Tests 

 - Unit test

Code changes

- None

Side effects

 - Possible performance regression

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: distsql/distsql.go [hunk 1]
```diff
@@ -47,24 +47,26 @@ func Select(ctx context.Context, sctx sessionctx.Context, kvReq *kv.Request, fie
...
+	}
+
 	// kvReq.MemTracker is used to trace and control memory usage in DistSQL layer;
 	// for streamResult, since it is a pipeline which has no buffer, it's not necessary to trace it;
 	// for selectResult, we just use the kvReq.MemTracker prepared for co-processor
 	// instead of creating a new one for simplification.
 	if kvReq.Streaming {
 		return &streamResult{
+			label:      "dag-stream",
+			sqlType:    label,
```
Comment (zz-jason) [inline distsql/distsql.go]: Seems we only need `label`? `sqlType` duplicates the information of `label`.

File: distsql/select_result.go [hunk 2]
```diff
@@ -96,6 +96,7 @@ func (r *selectResult) fetch(ctx context.Context) {
 
 		close(r.results)
 		duration := time.Since(startTime)
+		// TODO: Add a label to distinguish between success or failure
```
Comment (zz-jason) [inline distsql/select_result.go]: how about filing a github issue for this TODO action?

File: distsql/stream.go [hunk 3]
```diff
@@ -64,7 +68,10 @@ func (r *streamResult) Next(ctx context.Context, chk *chunk.Chunk) error {
 
 // readDataFromResponse read the data to result. Returns true means the resp is finished.
 func (r *streamResult) readDataFromResponse(ctx context.Context, resp kv.Response, result *tipb.Chunk) (bool, error) {
+	startTime := time.Now()
 	resultSubset, err := resp.Next(ctx)
+	// TODO: Add a label to distinguish between success or failure
+	metrics.DistSQLQueryHistgram.WithLabelValues(r.label, r.sqlType).Observe(time.Since(startTime).Seconds())
```
Comment (zz-jason) [inline distsql/stream.go]: should be deferred?

File: distsql/distsql.go (hunk 1 above)
Comment (SunRunAway) [reply distsql/distsql.go]: Hmmm, they're different, 
`sqlType` distinguishes the difference between general SQL or internal SQL and `label` distinguishes the difference between non-streaming dag or streaming dag.

File: distsql/stream.go (hunk 3 above)
Comment (SunRunAway) [reply distsql/stream.go]: We should only gauge the time task data transferring from TiKV but not plus the time for data transactions.

File: distsql/select_result.go (hunk 2 above)
Comment (SunRunAway) [reply distsql/select_result.go]: https://github.com/pingcap/tidb/issues/11397

File: store/mockstore/mocktikv/rpc.go [hunk 4]
```diff
@@ -619,17 +619,20 @@ type RPCClient struct {
...
-	ch := make(chan *tikvrpc.Lease)
-	go tikvrpc.CheckStreamTimeoutLoop(ch)
+	ch := make(chan *tikvrpc.Lease, 1024)
+	done := make(chan struct{})
+	go tikvrpc.CheckStreamTimeoutLoop(ch, done)
 	return &RPCClient{
 		Cluster:       cluster,
 		MvccStore:     mvccStore,
 		streamTimeout: ch,
+		done:          done,
```
Comment (tiancaiamao) [inline store/mockstore/mocktikv/rpc.go]: Why add the `done` channel?

Comment (tiancaiamao) [reply store/mockstore/mocktikv/rpc.go]: Will the `CheckStreamTimeoutLoop` goroutine exit?

Comment (SunRunAway) [reply store/mockstore/mocktikv/rpc.go]: > Why add the `done` channel?

The old way would panic
1. Invoke `RPCClient.Close` to close channel `streamTimeout`
2. An inflight SenRequest send a message into channel `streamTimeout`

**Comment by SunRunAway** [comment]
@tiancaiamao @zz-jason PTAL again.

**Comment by tiancaiamao** [comment]
LGTM
I suggest making an integration test (run tidb with the configure set to true) for coprocessor streaming.

File: store/tikv/tikvrpc/tikvrpc.go [hunk 5]
```diff
@@ -746,22 +746,36 @@ func (resp *CopStreamResponse) Recv() (*coprocessor.Response, error) {
...
-				// This channel close means goroutine should return.
-				return
+		case <-done:
+		drainLoop:
+			// Try my best cleaning the channel to make SendRequest which is blocking by it continues.
+			for {
+				select {
+				case <-ch:
+				default:
+					break drainLoop
```
Comment (zz-jason) [inline store/tikv/tikvrpc/tikvrpc.go]: I think the outside `for` loop is enough?

Comment (SunRunAway) [reply store/tikv/tikvrpc/tikvrpc.go]: Consider a huge number of `SendRequest` is waiting for channel `ch`, it may be the last chance to help them finish.

**Pull request merged**